sdk/identity/Azure.Identity/src/MacosNativeMethods.cs

```diff
@@ -0,0 +1,191 @@
+
+        public static void CFRelease(IntPtr cfRef)
+        {
+            if (cfRef != IntPtr.Zero)
+            {
+                Imports.CFRelease(cfRef);
+            }
+        }
+
+        private static void ThrowIfError(int status)
+        {
+            if (status != SecStatusCodeSuccess)
+            {
+                throw new InvalidOperationException(GetErrorMessageString(status));
+            }
+        }
+
+        private static string GetErrorMessageString(int status)
+        {
+            IntPtr messagePtr = IntPtr.Zero;
+            try
+            {
+                messagePtr = Imports.SecCopyErrorMessageString(status, IntPtr.Zero);
+                return GetStringFromCFStringPtr(messagePtr);
+            }
+            catch
+            {
+                switch (status)
+                {
+                    case SecStatusCodeNoSuchKeychain:
+                        return $"The keychain does not exist. [0x{status:x}]";
+                    case SecStatusCodeInvalidKeychain:
+                        return $"The keychain is not valid. [0x{status:x}]";
+                    case SecStatusCodeAuthFailed:
+                        return $"Authorization/Authentication failed. [0x{status:x}]";
+                    case SecStatusCodeDuplicateItem:
+                        return $"The item already exists. [0x{status:x}]";
+                    case SecStatusCodeItemNotFound:
+                        return $"The item cannot be found. [0x{status:x}]";
+                    case SecStatusCodeInteractionNotAllowed:
+                        return $"Interaction with the Security Server is not allowed. [0x{status:x}]";
+                    case SecStatusCodeInteractionRequired:
+                        return $"User interaction is required. [0x{status:x}]";
+                    case SecStatusCodeNoSuchAttr:
+                        return $"The attribute does not exist. [0x{status:x}]";
+                    default:
+                        return $"Unknown error. [0x{status:x}]";
+                }
+            }
+            finally
+            {
+                CFRelease(messagePtr);
+            }
+        }
+
+        private static string GetStringFromCFStringPtr(IntPtr handle)
+        {
+            IntPtr stringPtr = IntPtr.Zero;
+            try
+            {
+                stringPtr = Imports.CFStringGetCharactersPtr(handle);
+
+                if (stringPtr == IntPtr.Zero)
+                {
+                    int length = Imports.CFStringGetLength (handle);
+                    var range = new MacosNativeMethods.CFRange(0, length);
+                    stringPtr = Marshal.AllocCoTaskMem(length * 2);
+                    Imports.CFStringGetCharacters(handle, range, stringPtr);
+                }
+
+                return Marshal.PtrToStringAuto(stringPtr, 0);
```

Comment:
Would returning an empty string be better here?

Reply:
That's actually should've been `Marshal.PtrToStringAuto(stringPtr, length);`. Great catch!

Reply:
Fixed